### ISSUE TITLE
The max_det parameter of YOLO26 does not work

### ISSUE THREAD

**Issue by ZLT-a-LL** [ultralytics/ultralytics#23337]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

The max_det parameter of YOLO26 does not work，Even if max_det is set to 5000, the maximum number of identifications is still 300

### Environment

(yolo11) PS E:\A_worker\W_XM\Light_train> yolo checks
Ultralytics 8.4.6  Python-3.12.12 torch-2.9.1+cu128 CUDA:0 (NVIDIA GeForce RTX 3050 Ti Laptop GPU, 8192MiB)
Setup complete  (20 CPUs, 31.7 GB RAM, 857.4/953.9 GB disk)

OS                     Windows-11-10.0.26100-SP0
Environment            Windows
Python                 3.12.12
Install                pip
Path                   E:\Miniconda\envs\yolo11\Lib\site-packages\ultralytics
RAM                    31.71 GB
Disk                   857.4/953.9 GB
CPU                    12th Gen Intel Core i7-12700H
CPU count              20
GPU                    NVIDIA GeForce RTX 3050 Ti Laptop GPU, 8192MiB
GPU count              1
CUDA                   12.8

numpy                   2.4.1>=1.23.0
matplotlib              3.10.8>=3.3.0
opencv-python           4.13.0.90>=4.6.0
pillow                  12.0.0>=7.1.2
pyyaml                  6.0.3>=5.3.1
requests                2.32.5>=2.23.0
scipy                   1.16.3>=1.4.1
torch                   2.9.1+cu128>=1.8.0
torch                   2.9.1+cu128!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision             0.24.1+cu128>=0.9.0
psutil                  7.1.3>=5.8.0
polars                  1.36.1>=0.20.0
ultralytics-thop        2.0.18>=2.0.18
(yolo11) PS E:\A_worker\W_XM\Light_train> 

### Minimal Reproducible Example

from ultralytics import YOLO
import torch
from tqdm import tqdm
import cv2  # 需要导入cv2来获取视频的总帧数

# 检查是否可以使用GPU
print(torch.cuda.is_available())

# 加载训练好的YOLO模型
model = YOLO(r"E:\A_worker\W_XM\yolo11_car_seg\models\car_type_v21_1920.pt",)
             # task='detect')

# 视频路径
video_path = r"E:\A_worker\W_XM\yolo11_car_seg\test_img\flow\1.mp4"

# 使用OpenCV获取视频的总帧数，用于初始化进度条
cap = cv2.VideoCapture(video_path)
total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
cap.release()  # 立即释放资源

# 初始化进度条
pbar = tqdm(total=total_frames, desc="Processing video", unit="frames")

# 预测视频
results = model.track(
    source=video_path,
    save=True,
    save_txt=False,
    show=True,
    imgsz=1920,
    # iou=0.2,
    classes=[1],
    verbose=False,
    # line_width=2,
    # show_labels=False,
    # conf=0.5,
    tracker=r'E:\A_worker\W_XM\yolo11_car_seg\botsort_customs.yaml',
    # persist=True,
    stream=True  # 使用流式处理以避免内存问题
)

# 处理每一帧并更新进度条
for result in results:
    pbar.update(1)  # 每处理完一帧，进度条更新

# 关闭进度条
pbar.close()

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ZLT-a-LL, thank you for your interest in Ultralytics 🚀! This is an automated response to help us triage your issue quickly—an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. You’ve shared code (thanks!), and to make it a true MRE we’d also need a small, self-contained repro that others can run end-to-end without local paths. For example, please include:
-Exact `yolo` command or exact Python call where `max_det` is set (and where you expect it to apply)
-The full `model.track(...)` invocation including the `max_det` argument you’re using
-A short sample video or a tiny snippet that reproduces the behavior, plus the exact observed output (what shows “300” and where)
-Your `botsort_customs.yaml` contents if it may affect postprocessing/tracking
-Confirmation of whether the behavior happens in `predict()` as well as `track()` (same inputs/settings)

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by ZLT-a-LL** [comment]
The code above was sent by mistake. This is the test code：
from ultralytics import YOLO
import torch
from tqdm import tqdm
import cv2


print(torch.cuda.is_available())


model = YOLO(r"E:\A_worker\W_XM\best.pt",)
             # task='detect')


video_path = r"E:\A_worker\W_XM\20260120.mp4"


cap = cv2.VideoCapture(video_path)
total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
cap.release()


pbar = tqdm(total=total_frames, desc="Processing video", unit="frames")

results = model.predict(
    source=video_path,
    save=False,
    save_txt=False,
    show=True,
    imgsz=1920,
    # iou=0.2,
    # classes=[1],
    verbose=True,
    max_det=2000,
    # line_width=2,
    # show_labels=False,
    # conf=0.5,
    # tracker=r'E:\A_worker\W_XM\yolo11_car_seg\botsort_customs.yaml',
    # persist=True,
    stream=True
)

for result in results:
    pbar.update(1)

pbar.close()

**Comment by Y-T-G** [comment]
You can set it by modifying head

```
model.model.model[-1].max_det = 2000
```

**Comment by ZLT-a-LL** [comment]
Okay, thank you. Will the bug in the max_det parameter be fixed in the future? Also, I just tried adding the max_det parameter when converting to TensorRT, and it worked normally

**Comment by Y-T-G** [comment]
It should work for export. But prediction and validation currently doesn't work

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in the linked PR: [Support `max_det` and `agnostic_nms` for end2end (ultralytics 8.4.8)](https://github.com/ultralytics/ultralytics/pull/23396).

### What changed (and why it may fix this)
This PR improves end-to-end (NMS-free) inference behavior by ensuring `max_det` and `agnostic_nms` are actually applied when `end2end=True`, rather than being ignored in parts of the pipeline. Key updates include:

- Added a safe, generic `set_head_attr(**kwargs)` helper to set head (last-layer) attributes like `end2end`, `max_det`, and `agnostic_nms`.
- When `end2end=True`, both **predict** and **val** now push `max_det` and `agnostic_nms` into the model head so the settings take effect as expected.
- Added `agnostic_nms` support to the `Detect` head (affecting the top-k selection logic used in end-to-end flows).
- Export pipeline now passes `agnostic_nms` into the exported head to better match Python inference behavior after export.
- Training/attribute setup sets `max_det` for end-to-end models.
- Docs/tests updated accordingly, and the package version is bumped **8.4.7 → 8.4.8**.

This should improve consistency across **predict**, **validate**, **train-time validation**, and **exported** deployments for end-to-end models (including YOLO26 and YOLOv10).

### Credits
Thanks to @Y-T-G for the PR, with contributions from @Laughing-q and @glenn-jocher.

### How to test the fix
- Upgrade to the version mentioned in the PR:
  - `pip install -U ultralytics>=8.4.8`
- Or test the latest main branch directly:
  - `pip install git+https://github.com/ultralytics/ultralytics.git@main`
- Alternatively, waiting for the next official PyPI release is an option if `8.4.8` is not yet available in the environment being used.

### Feedback requested
Any confirmation on whether these changes resolve the issue on the affected setup would be really helpful. If the issue persists, sharing a minimal repro (model, args used like `end2end/max_det/agnostic_nms`, and the observed vs expected behavior) will help narrow it down further.

Thanks 🙏 for reporting this—further feedback is welcome if anything still looks off.
